config/configManager.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright The Ratify Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"os"
+	"sync"
+	"time"
+
+	ef "github.com/deislabs/ratify/pkg/executor/core"
+	"github.com/deislabs/ratify/pkg/policyprovider"
+	pf "github.com/deislabs/ratify/pkg/policyprovider/factory"
+	"github.com/deislabs/ratify/pkg/referrerstore"
+	sf "github.com/deislabs/ratify/pkg/referrerstore/factory"
+	"github.com/deislabs/ratify/pkg/verifier"
+	vf "github.com/deislabs/ratify/pkg/verifier/factory"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	configHash string
+)
+
+// Create a executor from configurationFile and setup config file watcher
+func GetExecutorAndWatchForUpdate(configFilePath string) (*ef.Executor, error) {
+	cf, err := Load(configFilePath)
+
+	if err != nil {
+		return &ef.Executor{}, err
+	}
+
+	configHash = cf.FileHash
+
+	stores, verifiers, policyEnforcer, err := createFromConfig(cf)
+
+	if err != nil {
+		return &ef.Executor{}, err
+	}
+
+	executor := ef.Executor{
+		Verifiers:      verifiers,
+		ReferrerStores: stores,
+		PolicyEnforcer: policyEnforcer,
+		Config:         &cf.ExecutorConfig,
+		Mu:             sync.RWMutex{},
```

Comment:
Is there a way to implement this lock free (read/write)? An option I would like to propose for evaluation is to use function type that will gets an executor object which  will be used to verify the subject something like 
```go
type GetExecutor func() *ef.Executor

var (
	configHash string
        executor  ef.Executor
)


GetExecutorAndWatchForUpdate(configFilePath string) (GetExecutor, error) {

    executor := := ef.Executor{
		Verifiers:      verifiers,
		ReferrerStores: stores,
.....
...
...

return func()  *ef.Executor { return &executor}, nil
}

in reloadExecutor() line 117 
if configHash != cf.FileHash { 
  create a new exceutor with new store, verifier and policy and replace the local executor to the new executor
  executor = newExecutor  
}

// on the usage side in handler.go https://github.com/deislabs/ratify/blob/0bd2d8c1a68912f59211681a5f53703531e0d10c/httpserver/handlers.go#L58 

executor := server.GetExecutor()
result, err := server.executor.VerifySubject(ctx, verifyParameters)  This ensures that new executor handle is fetched for every request and the in flight request will use the existing reference of the executor
```

Reply:
thanks so much for suggesting this :) , please review the updated the implementation if let me know if it aligns with your proposal

Reply:
Thanks @susanshi  for considering it. I hope it was simple without requiring any locks!